llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp

```diff
@@ -3604,7 +3613,11 @@ SDValue DAGTypeLegalizer::SoftPromoteHalfRes_UnaryOp(SDNode *N) {
   // Promote to the larger FP type.
   Op = DAG.getNode(GetPromotionOpcode(OVT, NVT), dl, NVT, Op);
 
-  SDValue Res = DAG.getNode(N->getOpcode(), dl, NVT, Op);
+  SDValue Res;
+  if (N->getOpcode() == ISD::AssertNoFPClass)
+    Res = DAG.getNode(N->getOpcode(), dl, NVT, Op, N->getOperand(1));
+  else
+    Res = DAG.getNode(N->getOpcode(), dl, NVT, Op);
```

Comment:
Treat as a separate case. You also can't preserve most of these fields on promotion, so maybe should just drop it